core/src/main/java/com/linecorp/armeria/internal/common/util/TargetLengthBasedClassNameAbbreviator.java

```diff
@@ -0,0 +1,115 @@
+/*
+ *  Copyright 2021 LINE Corporation
+ *
+ *  LINE Corporation licenses this file to you under the Apache License,
+ *  version 2.0 (the "License"); you may not use this file except in compliance
+ *  with the License. You may obtain a copy of the License at:
+ *
+ *    https://www.apache.org/licenses/LICENSE-2.0
+ *
+ *  Unless required by applicable law or agreed to in writing, software
+ *  distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ *  WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ *  License for the specific language governing permissions and limitations
+ *  under the License.
+ */
+/*
+ * Logback: the reliable, generic, fast and flexible logging framework.
+ * Copyright (C) 1999-2015, QOS.ch. All rights reserved.
+ *
+ * This program and the accompanying materials are dual-licensed under
+ * either the terms of the Eclipse Public License v1.0 as published by
+ * the Eclipse Foundation
+ *
+ *   or (per the licensee's choosing)
+ *
+ * under the terms of the GNU Lesser General Public License version 2.1
+ * as published by the Free Software Foundation.
+ */
+package com.linecorp.armeria.internal.common.util;
+
+public final class TargetLengthBasedClassNameAbbreviator {
+
+    private static final int MAX_DOTS = 16;
+    private final int targetLength;
+
+    static int computeDotIndexes(final String className, int[] dotArray) {
+        int dotCount = 0;
+        int k = 0;
+        while (true) {
+            // ignore the $ separator in our computations. This is both convenient
+            // and sensible.
+            k = className.indexOf('.', k);
+            if (k != -1 && dotCount < MAX_DOTS) {
+                dotArray[dotCount] = k;
+                dotCount++;
+                k++;
+            } else {
+                break;
+            }
+        }
+        return dotCount;
+    }
+
+    public TargetLengthBasedClassNameAbbreviator(int targetLength) {
+        this.targetLength = targetLength;
+    }
+
+    public String abbreviate(String fqClassName) {
+        final StringBuilder buf = new StringBuilder(targetLength);
```

Comment:
nit: Could you `TemporaryThreadLocals.get().stringBuilder()`?

Reply:
Thanks for recommending. I understand the purpose, but I am not sure that there will be no usage in same thread local. For example, from [`DefaultServiceRequestContext.toStringSlow()`](https://github.com/line/armeria/blob/master/core/src/main/java/com/linecorp/armeria/server/DefaultServiceRequestContext.java#L463), `MulticodeEncoder` or something else. Despite of cached abbreviation, is there possibility to be polluted? If wrong, please let me know. Or calling `buf.setLength(0)` after `buf = TemporaryThreadLocals.get().stringBuilder()` is okay?

Reply:
That's a good point. I think it's probably not worth a risk here because it's cached anyway.

Reply:
> is there possibility to be polluted?

If a method which uses `TemporaryThreadLocals.get().stringBuilder()` is called within a scope of `TemporaryThreadLocals.get().stringBuilder()`,
that must be a pollution.
For example:
```java
class Foo {
   String toString() {
       StringBuilder sb = TemporaryThreadLocals.get().stringBuilder();
       sb.append(...);
   }
}

class Bar {
    int value;
    Foo foo;
    String toString() {
        StringBuilder sb = TemporaryThreadLocals.get().stringBuilder();
        sb.append(value);
        // foo.toString() will pullute the `sb` of this method.
        sb.append(foo.toString());
        ...
    }
}
```